### DISCUSSION TITLE
Getting PnPContext via SharePoint Add-in Client Id and Secret/Using Sites.Selected

### DISCUSSION THREAD

**Discussion by tbithell** [pnp/pnpcore#474] [Q&A]
I've gone through the directions and have this working.  Unfortunately, more will have to happen in order to be able to use it in this project I'm on.

Is there a way to get PnPContext with a SPO add-in?  
I'm able to get ClientContext via the authprovider in Core Framework, but can't seem to sort out how to get PnP context with it.  

From our organization's perspective, .All permissions are not happening.  Site.Selected is real promising, but as best I can tell the SDK doesn't yet support it.  

I really want to use the SDK, so any help getting past this will be greatly appreciated. 

Thanks!

**Comment by jansenbe** [comment]
Hi @tbithell , the site.selected permissions are currently only working for Microsoft Graph only (so no REST or CSOM queries) and they only work in app-only context, hence you can't use PnP Core SDK with just those permissions. If that's what you need you'll need to limit yourselves to what Graph offers and use that. 

To get a PnPContext from an existing CSOM ClientContext you can use this: https://pnp.github.io/pnpcore/using-the-sdk/basics-pnpframework.html#using-pnp-core-sdk-when-pnp-framework-was-already-configured. Might not yet be available in 1.1, but it's in our nightly and we plan to release 1.2 very soon.

**Reply by tbithell** [reply]
Much appreciated!

Given that I can use the CSOM context to get PnP context that takes care of any needs as far as the rest api is concerned.  Are there any plans for app context graph access?  Ideally, I'd like our team to be able to have one path to SPO, that being the SDK.  I can say with confidence that enterprise use cases are going to be limited if .all permissions are the only route to using this thing for graph.

**Reply by jansenbe** [reply]
Not sure if I fully understand your question but:

- app only access today works for SPO REST, CSOM and Microsoft Graph
- app only comes in two flavors: the old ACS which works with REST/CSOM and via Azure AD which works for REST/CSOM and Microsoft Graph. 
  - ACS can deliver app-only scoped to a site collection, Azure AD app-only can't by default
  - .Selected permissions are the alternative for scoping an app's permissions to a site, but currently this only works when using Microsoft Graph and only for app-only scenarios. (https://developer.microsoft.com/en-us/graph/blogs/controlling-app-access-on-specific-sharepoint-site-collections/) 

PnP Core SDK will **always** use SPO REST, so if you want a 100% Microsoft Graph solution you need to use the Graph .NET SDK. Given you're looking at CSOM/PnP Framework I assume just using Microsoft Graph is not sufficient for your use case.
